### PR TITLE
Protocol buffers

### LINKED ISSUE #7: Galois Keys not in Payload serialization
After [PR6](https://github.com/OpenMined/PIR/pull/6), we'll introduce Galois Keys into the Payload, but not in the serialization. We need to add that to the serialization.

### LINKED ISSUE #10: Add protocol buffers to PIR
As we've discussed via Slack, the PSI team is going towards Protocol Buffers for serialization to better cross-language support (i.e. rather than depending on C++ first and then exporting serialization to other languages). We should update PIR to match, which means changing the existing json serialization to Protocol Buffers.

I've actually already implemented some of this in my own repo when I was experimenting with SealPIR. See the [rpc branch](https://github.com/kshehata/SealPIR/tree/rpc).

### PR DESCRIPTION
**Pull request by bcebere**
# Pull Request

## Description
Add GaloisKeys to Payload serialization
Use protobuffers for serialization.
fixes https://github.com/OpenMined/PIR/issues/7
fixes https://github.com/OpenMined/PIR/issues/10

## Affected Dependencies
List any dependencies that are required for this change.

## Type of Change
Please mark options that are relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation (non-breaking change which adds documentation)
- [x] Improvement (non-breaking change that improves the performance or reliability of existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I did follow the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md)
- [x] I did follow the [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have added tests for my changes

## Additional Context
Please also include relevant motivation and context.

### CODE REVIEW

**Comment by tlepoint** [comment]
Just food for thoughts (feel free to ignore this :)). Are you committed into adding Galois keys to each PIR query? Let's take for example [Epione](https://arxiv.org/pdf/2004.13293.pdf) which does PIR for contact tracing. In Sec 7.2.1, you can see that the client is sending 1120 queries; but they should /not/ send 1120 times the Galois keys. Similarly, I'm sure one could think of applications (including that of Epione) where the keys could be reused between sessions.

This will add logic on the server side, of course, but this also invites at splitting the "PIR query" from the "Galois keys upload".

File: pir/cpp/payload.proto [hunk 1]
```diff
@@ -0,0 +1,13 @@
+syntax = "proto2";
```
Comment (kshehata) [inline pir/cpp/payload.proto]: The proto should probably be in a separate directory to cpp. How about pir/proto ?

File: pir/cpp/payload.proto [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
```
Comment (kshehata) [inline pir/cpp/payload.proto]: A few other suggestions for this file:

1. Let's use proto3, unless there's a specific reason not to?
2. I'd suggest a comment at the top explaining the purpose of these protos, along with the boilerplate header

File: pir/cpp/payload.proto [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+package pir;
+
+message Payload {
```
Comment (kshehata) [inline pir/cpp/payload.proto]: Payload is somewhat generic. Perhaps something like PIRResponse or Reply? Also: please add a comment describing the purpose of this message.

File: pir/cpp/payload.proto [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  repeated bytes query = 1;
```
Comment (kshehata) [inline pir/cpp/payload.proto]: Is this the response to a query or the query itself? If the former, then this should be something like result or reply. Please also add a description of what this field is.

File: pir/cpp/payload.proto [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+message SessionPayload {
```
Comment (kshehata) [inline pir/cpp/payload.proto]: What do you mean by SessionPayload? Is this a query for a session value? It looks like a query to me. Perhaps this should be PIRQuery? Please add a comment description as well.

File: pir/cpp/payload.proto [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  required Payload query = 1;
```
Comment (kshehata) [inline pir/cpp/payload.proto]: Avoid required fields whenever possible. If we ever change the format of the query it'll be impossible to change. Please also add a description of this field.

File: pir/cpp/payload.proto [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  required int64 id = 2;
```
Comment (kshehata) [inline pir/cpp/payload.proto]: Avoid required fields. Also: ID of what?

File: pir/cpp/payload.proto [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  optional bytes galoiskeys = 3;
```
Comment (kshehata) [inline pir/cpp/payload.proto]: Please add a description to this field.

**Review by kshehata — CHANGES_REQUESTED** [review]
Glad to see this moving forward! I've made some comments on the proto file, since that's the most critical at this stage.

File: pir/cpp/payload.proto (hunk 5 above)
Comment (bcebere) [reply pir/cpp/payload.proto]: Payload is used for both query and response.

The session name comes from optimizing the GaloisKey roundtrips.

File: pir/cpp/payload.proto (hunk 2 above)
Comment (bcebere) [reply pir/cpp/payload.proto]: How can I set keys field to optional in proto3?

File: pir/cpp/payload.proto (hunk 7 above)
Comment (bcebere) [reply pir/cpp/payload.proto]: ID is session ID

File: pir/cpp/payload.proto (hunk 6 above)
Comment (bcebere) [reply pir/cpp/payload.proto]: That field is mandatory for the payload. Maybe another message is better if we break the interface.

File: pir/cpp/payload.proto (hunk 2 above)
Comment (kshehata) [reply pir/cpp/payload.proto]: Just don't set it to required.

File: pir/proto/payload.proto [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+syntax = "proto2";
+
+package pir;
+/*
+#include "pir/proto/pryload.pb.h"
```
Comment (kshehata) [inline pir/proto/payload.proto]: Remove?

File: pir/proto/payload.proto [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+*/
+
+/*PayloadData holds the request/response data buffers, usually ciphertexts */
```
Comment (kshehata) [inline pir/proto/payload.proto]: Please see here for how to comment proto files:
https://github.com/kshehata/SealPIR/blob/rpc/pir.proto

Comment (kshehata) [inline pir/proto/payload.proto]: This can be more specific. If this intended as a list of ciphertexts, then just call it a list of ciphertexts. That makes the message definitions clearer.

File: pir/proto/payload.proto [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+message PayloadData {
+  repeated bytes data = 1; //query or response data
```
Comment (kshehata) [inline pir/proto/payload.proto]: This is an antipattern. If you're naming a field in a proto just "data", that means it isn't specific to what you're doing and probably should be renamed.

File: pir/proto/payload.proto [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+/*Payload contains the entire request or response message */
+message Payload {
```
Comment (kshehata) [inline pir/proto/payload.proto]: As above, let's name this more specifically. What is the payload for? So far as I can tell, this is only ever used for a Request. So why not name it as such? The more direct the name, the clearer your intentions are.

File: pir/proto/payload.proto [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  required PayloadData data = 1; //payload data
+  required int64 id = 2; //session ID
+  optional bytes galoiskeys = 3;//optional Galois keys
+}
```
Comment (kshehata) [inline pir/proto/payload.proto:39]: We also need a message for the Server reply. It might have just one field, but that's ok. That makes the message structure more adaptable.

File: pir/proto/payload.proto [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+*/
+
+/*PayloadData holds the request/response data buffers, usually ciphertexts */
+message PayloadData {
+  repeated bytes data = 1; //query or response data
+}
+
+/*Payload contains the entire request or response message */
+message Payload {
+  required PayloadData data = 1; //payload data
```
Comment (kshehata) [inline pir/proto/payload.proto]: Trust me, you do _not_ want to set required here. From https://developers.google.com/protocol-buffers/docs/proto 

> Required Is Forever You should be very careful about marking fields as required. If at some point you wish to stop writing or sending a required field, it will be problematic to change the field to an optional field – old readers will consider messages without this field to be incomplete and may reject or drop them unintentionally. You should consider writing application-specific custom validation routines for your buffers instead. Some engineers at Google have come to the conclusion that using required does more harm than good; they prefer to use only optional and repeated. However, this view is not universal.

From my point of view, required is basically _never_ needed and usually does more harm than good. It's more important to validate messages you receive than try to have the library do that for you, since any change will either break backwards compatibility or break your validation. This is a lesson learned after building _many_ systems at Google.

**Review by kshehata — CHANGES_REQUESTED** [review]

File: pir/proto/payload.proto (hunk 13 above)
Comment (bcebere) [reply pir/proto/payload.proto:39]: This is for the server too

File: pir/cpp/client.h [hunk 15]
```diff
@@ -41,14 +41,14 @@ class PIRClient {
...
-  StatusOr<PIRPayload> CreateRequest(std::size_t /*index*/) const;
+  StatusOr<PIRQuery> CreateRequest(std::size_t /*index*/) const;
 
   /**
    * Extracts server response
    * @param[in] response Server output
    * @returns InvalidArgument if the decryption fails
    **/
-  StatusOr<int64_t> ProcessResponse(const PIRPayload& response) const;
+  StatusOr<int64_t> ProcessResponse(const PIRReply& response);
```
Comment (kshehata) [inline pir/cpp/client.h]: Why is this no longer const?

File: pir/cpp/payload.h [hunk 16]
```diff
@@ -30,40 +31,99 @@ using seal::GaloisKeys;
 using std::optional;
 using buff_type = std::vector<seal::Ciphertext>;
 
-class PIRPayload {
+class PIRCiphertexts {
```
Comment (kshehata) [inline pir/cpp/payload.h]: These classes seem to largely duplicate the functionality of the protocol buffers. How about just a utility function to extract the ciphertexts, rather than defining a set of mirror classes?

**Review by kshehata — CHANGES_REQUESTED** [review]
Definitely liking the protos now! I'd suggest we use those directly as what gets passed around, instead of making our own classes. That reduces code duplication and makes it easier to change things in the future. It does mean that it forces serialization of the ciphertexts into the PB objects, but not the serialization of the PB itself.

File: pir/cpp/client.h (hunk 15 above)
Comment (bcebere) [reply pir/cpp/client.h]: Leftovers from sessionID experiment

File: pir/cpp/payload.h [hunk 17]
```diff
@@ -17,55 +17,70 @@
...
+ * @param[in] The encoded ciphertext.
+ * @returns InvalidArgument if the decoding fails.
+ **/
+StatusOr<buff_type> LoadCiphertexts(
+    const std::shared_ptr<seal::SEALContext>& ctx, const Ciphertexts& encoded);
+/**
+ * Saves the Ciphertexts to a protobuffer.
+ * @returns InvalidArgument if the encoding fails
+ **/
+StatusOr<Ciphertexts> SaveCiphertexts(const buff_type& buff);
```
Comment (kshehata) [inline pir/cpp/payload.h]: Same here

File: pir/cpp/payload.h [hunk 18]
```diff
@@ -17,55 +17,70 @@
...
 
-  const optional<GaloisKeys>& GetKeys() const { return keys_; }
+/**
+ * Saves a SEAL object to a string.
```
Comment (kshehata) [inline pir/cpp/payload.h]: Please add a little more detail here, including which types this is intended for and when it can be used (i.e. Ciphertext and GaloisKeys)

File: pir/cpp/payload.h [hunk 19]
```diff
@@ -17,55 +17,70 @@
...
+  std::stringstream stream;
 
- private:
-  PIRPayload(const std::vector<seal::Ciphertext>& buff,
-             const optional<GaloisKeys>& keys = {})
-      : buff_(buff), keys_(keys){};
+  try {
+    sealobj.save(stream);
+  } catch (const std::exception& e) {
+    return InvalidArgumentError(e.what());
```
Comment (kshehata) [inline pir/cpp/payload.h]: Is this really an invalid argument, or more likely an InternalError?

File: pir/cpp/payload.h [hunk 20]
```diff
@@ -17,55 +17,70 @@
...
-   **/
-  const buff_type& Get() const;
-  PIRPayload() = delete;
+/**
+ * Decodes and loads a PIR Ciphertext.
+ * @param[in] The SEAL context, for buffer allocations.
+ * @param[in] The encoded ciphertext.
+ * @returns InvalidArgument if the decoding fails.
+ **/
+StatusOr<buff_type> LoadCiphertexts(
```
Comment (kshehata) [inline pir/cpp/payload.h]: The buff_type here actually makes this harder to read. Perhaps just use the vector<Ciphertext> to make this more obvious?

File: pir/cpp/payload.cpp [hunk 21]
```diff
@@ -31,89 +27,26 @@ using ::private_join_and_compute::StatusOr;
...
 
-StatusOr<std::string> PIRPayload::Save() {
-  std::vector<std::string> interm(buff_.size());
-
-  for (size_t idx = 0; idx < buff_.size(); ++idx) {
-    ASSIGN_OR_RETURN(interm[idx], serializeCT(buff_[idx]));
+StatusOr<Ciphertexts> SaveCiphertexts(const buff_type& ct_) {
+  Ciphertexts output;
+  for (size_t idx = 0; idx < ct_.size(); ++idx) {
+    ASSIGN_OR_RETURN(auto ct, SEALSerialize<Ciphertext>(ct_[idx]));
```
Comment (kshehata) [inline pir/cpp/payload.cpp]: The name "ct" here is confusing. Could you change it to make it more obvious this is the string version?

File: pir/cpp/payload.cpp [hunk 22]
```diff
@@ -31,89 +27,26 @@ using ::private_join_and_compute::StatusOr;
...
+  return buff;
 }
 
-StatusOr<std::string> PIRPayload::Save() {
-  std::vector<std::string> interm(buff_.size());
-
-  for (size_t idx = 0; idx < buff_.size(); ++idx) {
-    ASSIGN_OR_RETURN(interm[idx], serializeCT(buff_[idx]));
+StatusOr<Ciphertexts> SaveCiphertexts(const buff_type& ct_) {
+  Ciphertexts output;
```
Comment (kshehata) [inline pir/cpp/payload.cpp]: Perhaps something in this name to indicate that it's a PB?

File: pir/cpp/server.h [hunk 23]
```diff
@@ -55,7 +55,7 @@ class PIRServer {
    * @returns InvalidArgument if the deserialization or encrypted operations
    *fail
    **/
-  StatusOr<PIRPayload> ProcessRequest(const PIRPayload& request) const;
+  StatusOr<Reply> ProcessRequest(const Query& request) const;
```
Comment (kshehata) [inline pir/cpp/server.h]: I've just noticed that we don't seem to be consistent with the use of the terms "query" and "request". How about we use "request" for the overall request from the client, and "query" for an individual query vector within the request? I think that'll make things easier. What do you think?

File: pir/cpp/server.cpp [hunk 24]
```diff
@@ -49,22 +50,26 @@ StatusOr<std::unique_ptr<PIRServer>> PIRServer::Create(
...
 
   ASSIGN_OR_RETURN(auto mult_results, db_->multiply(selection_vector));
 
   seal::Ciphertext result;
   context_->Evaluator()->add_many(mult_results, result);
 
-  return PIRPayload::Load(vector<seal::Ciphertext>{result});
+  Reply reply;
+  ASSIGN_OR_RETURN(*reply.mutable_reply(),
+                   SaveCiphertexts(vector<seal::Ciphertext>{result}));
```
Comment (kshehata) [inline pir/cpp/server.cpp]: If this is our use case, perhaps it would be better to pass into SaveCiphertexts a pointer to the proto into which to save output? Seems like it would be more readable and more efficient.

File: pir/cpp/client.h [hunk 25]
```diff
@@ -41,14 +41,14 @@ class PIRClient {
...
-  StatusOr<PIRPayload> CreateRequest(std::size_t /*index*/) const;
+  StatusOr<Query> CreateRequest(std::size_t /*index*/) const;
 
   /**
    * Extracts server response
    * @param[in] response Server output
    * @returns InvalidArgument if the decryption fails
    **/
-  StatusOr<int64_t> ProcessResponse(const PIRPayload& response) const;
+  StatusOr<int64_t> ProcessResponse(const Reply& response) const;
```
Comment (kshehata) [inline pir/cpp/client.h]: Also noticed that we're not consistent with "response" vs "reply". Perhaps in this case we can just use "response"?

File: pir/cpp/client.cpp [hunk 26]
```diff
@@ -94,20 +95,31 @@ StatusOr<PIRPayload> PIRClient::CreateRequest(std::size_t desired_index) const {
   try {
     gal_keys =
         keygen_->galois_keys_local(generate_galois_elts(poly_modulus_degree));
+
   } catch (const std::exception& e) {
     return InternalError(e.what());
   }
-  return PIRPayload::Load(query, gal_keys);
+
+  ASSIGN_OR_RETURN(auto encoded_query, SaveCiphertexts(query));
```
Comment (kshehata) [inline pir/cpp/client.cpp]: You should be able to create the proto and then save directly to it. That would be both more efficient and easier to read. I'd also suggest we change the serialization methods so they take pointers to where to store output.

File: pir/cpp/client_test.cpp [hunk 27]
```diff
@@ -49,10 +51,13 @@ class PIRClientTest : public ::testing::Test {
...
-  auto payload = client_->CreateRequest(desired_index).ValueOrDie();
+  auto req_proto = client_->CreateRequest(desired_index).ValueOrDie();
+  auto req =
+      LoadCiphertexts(Context()->SEALContext(), req_proto.query()).ValueOrDie();
+
   Plaintext pt;
-  ASSERT_EQ(payload.Get().size(), 1);
-  Decryptor()->decrypt(payload.Get()[0], pt);
+  ASSERT_EQ(req.size(), 1);
+  Decryptor()->decrypt(req[0], pt);
```
Comment (kshehata) [inline pir/cpp/client_test.cpp:60]: Ok, that's cool. I didn't know you could just straight decrypt a string!

File: pir/cpp/client_test.cpp [hunk 28]
```diff
@@ -72,12 +77,52 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
 
-  auto result = client_->ProcessResponse(payload).ValueOrDie();
+  Reply reply;
+  *reply.mutable_reply() = SaveCiphertexts(ct).ValueOrDie();
+
+  auto result = client_->ProcessResponse(reply).ValueOrDie();
   ASSERT_EQ(result, value);
 }
 
+TEST_F(PIRClientTest, TestPayloadSerialization) {
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: Why is this here instead of with payload?

File: pir/cpp/client_test.cpp [hunk 29]
```diff
@@ -72,12 +77,52 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
+  auto keygen_ = make_unique<KeyGenerator>(Context()->SEALContext());
+  auto elts = generate_galois_elts(DEFAULT_POLY_MODULUS_DEGREE);
+  GaloisKeys gal_keys = keygen_->galois_keys_local(elts);
+
+  auto encoded_query = SaveCiphertexts(ct).ValueOrDie();
+  auto encoded_keys = SEALSerialize<GaloisKeys>(gal_keys).ValueOrDie();
+
+  Query proto_query;
+  *proto_query.mutable_query() = encoded_query;
+  proto_query.set_keys(encoded_keys);
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: This would be a lot more readable if you create the query proto first and then assign into it directly.

File: pir/cpp/server_test.cpp [hunk 30]
```diff
@@ -84,6 +84,7 @@ class PIRServerTest : public ::testing::Test {
     evaluator_ = make_unique<Evaluator>(context);
     decryptor_ = make_unique<Decryptor>(context, keygen_->secret_key());
   }
+  PIRContext* Context() { return server_->Context(); }
```
Comment (kshehata) [inline pir/cpp/server_test.cpp]: Isn't this function already in Server?

**Review by kshehata — CHANGES_REQUESTED** [review]

File: pir/cpp/client_test.cpp (hunk 28 above)
Comment (bcebere) [reply pir/cpp/client_test.cpp]: Because the client does the serialization.

File: pir/cpp/payload.cpp [hunk 31]
```diff
@@ -31,89 +27,30 @@ using ::private_join_and_compute::StatusOr;
...
-  output.SetArray();
-  for (size_t idx = 0; idx < buff_.size(); ++idx) {
-    output.PushBack(
-        rapidjson::Value().SetString(interm[idx].data(), interm[idx].size(),
-                                     output.GetAllocator()),
-        output.GetAllocator());
+  for (size_t idx = 0; idx < ciphertexts.size(); ++idx) {
+    ASSIGN_OR_RETURN(auto ciphertext_str,
+                     SEALSerialize<Ciphertext>(ciphertexts[idx]));
+    output->add_ct(ciphertext_str);
```
Comment (kshehata) [inline pir/cpp/payload.cpp]: Does this make a second copy? Is it possible to do this without making any copies?

File: pir/cpp/payload.h [hunk 32]
```diff
@@ -17,55 +17,76 @@
 #ifndef PIR_PAYLOAD_H_
 #define PIR_PAYLOAD_H_
 
-#include <optional>
```
Comment (kshehata) [inline pir/cpp/payload.h:20]: (More a note for myself) does it still make sense to have these function in payload.h? These are really just serializing / deserializing SEAL objects. Thoughts?

File: pir/cpp/client.cpp [hunk 33]
```diff
@@ -94,20 +95,31 @@ StatusOr<PIRPayload> PIRClient::CreateRequest(std::size_t desired_index) const {
   try {
     gal_keys =
         keygen_->galois_keys_local(generate_galois_elts(poly_modulus_degree));
+
```
Comment (kshehata) [inline pir/cpp/client.cpp]: Nit: please avoid whitespace changes like this unless they're important for readability. This one will cause a conflict when I go to merge my PR :)

File: pir/cpp/client.cpp [hunk 34]
```diff
@@ -94,20 +95,31 @@ StatusOr<PIRPayload> PIRClient::CreateRequest(std::size_t desired_index) const {
...
   } catch (const std::exception& e) {
     return InternalError(e.what());
   }
-  return PIRPayload::Load(query, gal_keys);
+
+  Request proto_query;
```
Comment (kshehata) [inline pir/cpp/client.cpp]: Perhaps request_proto is a better name here?

File: pir/cpp/client.cpp [hunk 35]
```diff
@@ -94,20 +95,31 @@ StatusOr<PIRPayload> PIRClient::CreateRequest(std::size_t desired_index) const {
...
+
+  RETURN_IF_ERROR(SaveCiphertexts(query, proto_query.mutable_query()));
+  ASSIGN_OR_RETURN(auto encoded_keys, SEALSerialize<GaloisKeys>(gal_keys));
```
Comment (kshehata) [inline pir/cpp/client.cpp]: Similar to SaveCiphertexts, if the primary use case for SEALSerialize is for use with a PB, perhaps it would be more efficient and easier to use for it to take a pointer to a mutable string?

File: pir/cpp/server.cpp [hunk 36]
```diff
@@ -49,22 +50,27 @@ StatusOr<std::unique_ptr<PIRServer>> PIRServer::Create(
...
 }
 
-StatusOr<PIRPayload> PIRServer::ProcessRequest(
-    const PIRPayload& payload) const {
-  if (!payload.GetKeys()) {
-    return InvalidArgumentError("Must have Galois Keys in request");
-  }
+StatusOr<Response> PIRServer::ProcessRequest(
+    const Request& request_proto) const {
+  ASSIGN_OR_RETURN(auto request, LoadCiphertexts(context_->SEALContext(),
```
Comment (kshehata) [inline pir/cpp/server.cpp]: Perhaps rename this to query to make it obvious the difference between this request_proto?

File: pir/cpp/client_test.cpp [hunk 37]
```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
   Context()->Encoder()->encode(value, pt);
   vector<Ciphertext> ct(1);
   Encryptor()->encrypt(pt, ct[0]);
-  PIRPayload payload = PIRPayload::Load(ct);
 
-  auto result = client_->ProcessResponse(payload).ValueOrDie();
+  Response reply;
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: Please name this response for consistency

File: pir/cpp/client_test.cpp [hunk 38]
```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
+  SaveCiphertexts(ct, reply.mutable_reply());
+
+  auto result = client_->ProcessResponse(reply).ValueOrDie();
   ASSERT_EQ(result, value);
 }
 
+TEST_F(PIRClientTest, TestRequestSerialization) {
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: Shouldn't this test be under payload_test.cpp?

File: pir/cpp/client_test.cpp [hunk 39]
```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
+  int64_t value = 987654321;
+  Plaintext pt;
+  Context()->Encoder()->encode(value, pt);
+  vector<Ciphertext> ct(1);
+  Encryptor()->encrypt(pt, ct[0]);
+
+  Response reply_proto;
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: If this is testing request serialization, why is this a response proto? Shouldn't this be a separate test?

File: pir/cpp/client_test.cpp [hunk 40]
```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
+  SaveCiphertexts(ct, reply_proto.mutable_reply());
+
+  auto reloaded = LoadCiphertexts(Context()->SEALContext(), reply_proto.reply())
+                      .ValueOrDie();
+
+  ASSERT_EQ(reloaded.size(), 1);
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: Perhaps check that the CTs are the same?

File: pir/cpp/client_test.cpp [hunk 41]
```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
+  Plaintext pt;
+  Context()->Encoder()->encode(value, pt);
+  vector<Ciphertext> ct(1);
+  Encryptor()->encrypt(pt, ct[0]);
+
+  Response reply_proto;
+  SaveCiphertexts(ct, reply_proto.mutable_reply());
+
+  auto reloaded = LoadCiphertexts(Context()->SEALContext(), reply_proto.reply())
+                      .ValueOrDie();
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: Rather than doing ValueOrDie, it's a lot easier to debug the test if you do an assert on status being ok and output the status to_string value.

File: pir/cpp/client_test.cpp [hunk 42]
```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
+
+  ASSERT_EQ(reloaded.size(), 1);
+
+  auto keygen_ = make_unique<KeyGenerator>(Context()->SEALContext());
+  auto elts = generate_galois_elts(DEFAULT_POLY_MODULUS_DEGREE);
+  GaloisKeys gal_keys = keygen_->galois_keys_local(elts);
+
+  Request proto_query;
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: Call this request for consistency.

File: pir/cpp/client_test.cpp [hunk 43]
```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
+
+  auto request = LoadCiphertexts(Context()->SEALContext(), proto_query.query())
+                     .ValueOrDie();
+  auto keys =
+      SEALDeserialize<GaloisKeys>(Context()->SEALContext(), proto_query.keys())
+          .ValueOrDie();
+
+  ASSERT_EQ(request.size(), 1);
+  for (auto& elt : elts) {
+    ASSERT_TRUE(keys.has_key(elt));
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: There are much better asserts in gmock. I'd suggest using ElementsAreArray here.

File: pir/cpp/client_test.cpp [hunk 44]
```diff
@@ -72,12 +77,50 @@ TEST_F(PIRClientTest, TestProcessResponse) {
...
+  auto encoded_keys = SEALSerialize<GaloisKeys>(gal_keys).ValueOrDie();
+  proto_query.set_keys(encoded_keys);
+
+  auto request = LoadCiphertexts(Context()->SEALContext(), proto_query.query())
+                     .ValueOrDie();
+  auto keys =
+      SEALDeserialize<GaloisKeys>(Context()->SEALContext(), proto_query.keys())
+          .ValueOrDie();
+
+  ASSERT_EQ(request.size(), 1);
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: Check that CT matches as well?

File: pir/cpp/client_test.cpp [hunk 45]
```diff
@@ -101,10 +144,12 @@ TEST_P(CreateRequestTest, TestCreateRequest_MoreThanOneCT) {
   auto payload_or = client_->CreateRequest(desired_index);
   ASSERT_TRUE(payload_or.ok())
       << "Status is: " << payload_or.status().ToString();
-  auto payload = payload_or.ValueOrDie();
-  ASSERT_EQ(payload.Get().size(), dbsize / poly_modulus_degree + 1);
+  auto payload =
```
Comment (kshehata) [inline pir/cpp/client_test.cpp]: Rename this to query please, to be consistent with the new naming.

File: pir/cpp/server_test.cpp [hunk 46]
```diff
@@ -111,20 +111,27 @@ TEST_F(PIRServerTest, TestProcessRequest_SingleCT) {
   pt.set_zero();
   pt[desired_index] = 1;
 
-  vector<Ciphertext> query(1);
-  encryptor_->encrypt(pt, query[0]);
+  vector<Ciphertext> request(1);
+  encryptor_->encrypt(pt, request[0]);
```
Comment (kshehata) [inline pir/cpp/server_test.cpp]: Based on the new naming, isn't this a query?

File: pir/cpp/server_test.cpp [hunk 47]
```diff
@@ -111,20 +111,27 @@ TEST_F(PIRServerTest, TestProcessRequest_SingleCT) {
...
   GaloisKeys gal_keys =
       keygen_->galois_keys_local(generate_galois_elts(POLY_MODULUS_DEGREE));
-  auto payload = PIRPayload::Load(query, gal_keys);
 
-  auto result_or = server_->ProcessRequest(payload);
+  auto encoded_keys = SEALSerialize<GaloisKeys>(gal_keys).ValueOrDie();
+
+  Request proto_request;
+  SaveCiphertexts(request, proto_request.mutable_query());
+  proto_request.set_keys(encoded_keys);
```
Comment (kshehata) [inline pir/cpp/server_test.cpp]: Why not call SEALSerialize on this line?

File: pir/cpp/server_test.cpp [hunk 48]
```diff
@@ -136,22 +143,29 @@ TEST_F(PIRServerTest, TestProcessRequest_MultiCT) {
   Plaintext pt(POLY_MODULUS_DEGREE);
   pt.set_zero();
 
-  vector<Ciphertext> query(2);
-  encryptor_->encrypt(pt, query[0]);
+  vector<Ciphertext> request(2);
```
Comment (kshehata) [inline pir/cpp/server_test.cpp]: Same as above, is this not now a query based on our naming?

File: pir/cpp/server_test.cpp [hunk 49]
```diff
@@ -165,19 +179,27 @@ TEST_F(PIRServerTest, TestProcessRequestZeroInput) {
   Plaintext pt(POLY_MODULUS_DEGREE);
   pt.set_zero();
 
-  vector<Ciphertext> query(1);
-  encryptor_->encrypt(pt, query[0]);
+  vector<Ciphertext> request(1);
```
Comment (kshehata) [inline pir/cpp/server_test.cpp]: As above, call this query?

**Review by kshehata — CHANGES_REQUESTED** [review]
We're getting to the point that most of my comments are about tests. That's a good sign of progress! I think we've gotten through all of the structural changes, except possibly the types used for SEAL(De)Serialize. Let's discuss where the serialization functions should live; I'm not convinced payload is the right place for them any more.

File: pir/cpp/client_test.cpp (hunk 43 above)
Comment (bcebere) [reply pir/cpp/client_test.cpp]: Not sure how should I use ElementsAreArray in this particular scenario.

Comment (kshehata) [reply pir/cpp/client_test.cpp]: Could you not do something like:
`ASSERT_THAT(keys, ElementsAreArray(elts));`

File: pir/cpp/BUILD [hunk 50]
```diff
@@ -42,7 +42,7 @@ cc_library(
 
 cc_test(
     name = "pir_test",
-    srcs = ["client_test.cpp", "server_test.cpp"],
+    srcs = ["client_test.cpp", "server_test.cpp", "serialization.cpp"],
```
Comment (kshehata) [inline pir/cpp/BUILD]: serialization_test?

File: pir/cpp/serialization_test.cpp [hunk 51]
```diff
@@ -0,0 +1,101 @@
...
+using std::tuple;
+
+class PIRSerializationTest : public ::testing::Test {
+ protected:
+  static constexpr std::size_t DB_SIZE = 100;
+  void SetUp() { SetUpDB(DB_SIZE); }
+
+  void SetUpDB(size_t dbsize) {
+    pir_params_ = PIRParameters::Create(dbsize);
+    client_ = PIRClient::Create(pir_params_).ValueOrDie();
```
Comment (kshehata) [inline pir/cpp/serialization_test.cpp]: Why is a client needed here? If you only need the decryptor and encryptor, perhaps just make those directly?

**Review by kshehata — APPROVED** [review]
LGTM other than a couple of nits. Good work, this code looks great!
